dot/types/grandpa_test.go

```diff
@@ -7,55 +7,161 @@ import (
+		"05dfedfffa66f99f16fe8b938fd8df76a92225228a1ca075230b99a2d9e173c561952e1e378b701915ca188d2c832ef92a3fab8e455" +
+		"f32570c0807")
+	identity := common.MustHexToBytes("0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee")
+	identityPubKey, _ := ed25519.NewPublicKey(identity)
+	firstVote := GrandpaVote{
+		Hash:   common.MustHexToHash("0x4801b8e62d31167d30c893cc1970f6a0e289420282a4b245b75f2c46fb308af1"),
+		Number: uint32(10),
+	}
+	secondVote := GrandpaVote{
+		Hash:   common.MustHexToHash("0xc314327941fdd924bc67fd72651c40aececd485ca3e878c21e02abb40feae5bd"),
+		Number: uint32(10),
+	}
+
+	firstSignatureArray := mustHexTo64BArray(t, "0xd7292caacc62504365f179892a7399f233944bf261f8a3f66260f70e0016f2d"+
+		"b63922726b015c82dc7131f4730fbec61f71672a571453e51029bfb469070900f")
+
+	secondSignatureArray := mustHexTo64BArray(t, "0xb3c408b74905dfedfffa66f99f16fe8b938fd8df76a92225228a1ca07523"+
+		"0b99a2d9e173c561952e1e378b701915ca188d2c832ef92a3fab8e455f32570c0807")
+
+	var authorityID [32]byte
+	copy(authorityID[:], identityPubKey.Encode())
+
+	grandpaEquivocation := GrandpaEquivocation{
+		RoundNumber:     1,
+		ID:              authorityID,
+		FirstVote:       firstVote,
+		FirstSignature:  firstSignatureArray,
+		SecondVote:      secondVote,
+		SecondSignature: secondSignatureArray,
+	}
+	preVoteEquivocation := PreVoteEquivocation(grandpaEquivocation)
+	equivocationVote := NewGrandpaEquivocation()
+	err := equivocationVote.Set(preVoteEquivocation)
+	require.NoError(t, err)
+
+	equivocationProof := GrandpaEquivocationProof{
+		SetID:        1,
+		Equivocation: *equivocationVote,
+	}
+
+	actualEncoding := scale.MustMarshal(equivocationProof)
+	require.Equal(t, expectedEncoding, actualEncoding)
+}
+
+func Test_PreVoteEquivocation_ScaleCodec(t *testing.T) {
+	t.Parallel()
+	firstVote := GrandpaVote{
 		Hash:   common.Hash{0xa, 0xb, 0xc, 0xd},
 		Number: 999,
 	}
+	secondVote := GrandpaVote{
+		Hash:   common.Hash{0xd, 0xc, 0xb, 0xa},
+		Number: 999,
+	}
+	signature := [64]byte{1, 2, 3, 4}
+	keypair, err := ed25519.GenerateKeypair()
+	require.NoError(t, err)
+
+	var authorityID [32]byte
+	copy(authorityID[:], keypair.Public().Encode())
+
+	equivocation := GrandpaEquivocation{
+		RoundNumber:     0,
+		ID:              authorityID,
+		FirstVote:       firstVote,
+		FirstSignature:  signature,
+		SecondVote:      secondVote,
+		SecondSignature: signature,
+	}
 
-	enc, err := scale.Marshal(testVote)
+	equivPreVote := PreVoteEquivocation(equivocation)
+	equivVote := NewGrandpaEquivocation()
+	err = equivVote.Set(equivPreVote)
 	require.NoError(t, err)
-	require.Equal(t, exp, enc)
+	encoding := scale.MustMarshal(*equivVote)
 
-	dec := GrandpaVote{}
-	err = scale.Unmarshal(enc, &dec)
+	grandpaEquivocation := NewGrandpaEquivocation()
+	err = scale.Unmarshal(encoding, grandpaEquivocation)
 	require.NoError(t, err)
-	require.Equal(t, testVote, dec)
+	require.Equal(t, equivVote, grandpaEquivocation)
+}
```

Comment:
This test that after marshaling we unmarshal to the same struct. I could replace `equivVote` with any struct and this test would work.

Instead, we could test it against some other scale encoding. So we could get rust equivalent of  `equivVote`  and marshal it using rust scale library and test that
- our struct also marshals to the same value.
- we can unmarshal it to our struct

Reply:
Do you think since all this is already tested in an above test, I should just remove this one and test against anencoding above?

Reply:
I opted to delete this test since its already covered in above test